### PR TITLE
[SIS-082] search activity cards backend new

### LINKED ISSUE #82: [SIS-082] Search activity cards backend
> As a user of the Sistema Lesson Planner, I want to be able to search for activity cards.

## Technical Information

Endpoints to implement: 
`searchActivityCards`
`getActivityCardPreview`

### searchActivityCards
**Description:** Given some user input, generate a list of activity cards in the Sistema [Curriculum Materials](https://drive.google.com/drive/folders/1dki-Xc9XusQm86IvIjNC23Rv8rtQQTs-) folder. Cap results at Constants.maxSearch which is an int determining the max number of suggested activity cards per search. Place this number in `constants.js` file. 

**Discussion:** 

- Typing in the Add Activity Card search bar should generate a list of results for this. I would ping the designers for how many search results you recommend displaying (see [UI](https://www.figma.com/file/kSZduD0A5hk8sHmvqiC7DE/Sistema-Hi-Fidelity-Designs-%2B-Design-System?node-id=59%3A12896&t=VFmAxlq4E25Sw94q-0)) because 2 is too little. 
- Don't forget to search by the user's selected filters, which will be strings delimited by commas in the Drive file's description. We confirmed that this should automatically be counted in Drive's list but double check what happens when multiple filters are applied (enforce that filters are AND not OR). 
For Sistema, they edit tags here:
image

### getActivityCardPreview
**Description:** Get a preview of a requested activity card from Drive and display it [here](https://www.figma.com/file/kSZduD0A5hk8sHmvqiC7DE/Sistema-Hi-Fidelity-Designs-%2B-Design-System?node-id=59%3A13353&t=VFmAxlq4E25Sw94q-0). _Do not download the activity card to RNFS_. 

**Discussion:** 

- See [wiki](https://github.com/uoftblueprint/sistema/wiki/Search-API#getactivitycardpreviewid-string) and [Endpoint Reference](https://docs.google.com/document/d/1JdqTkQoUPN7ov9fGmf_vruavwfDfUzUdNY5I9AccnKQ/edit#) for prior discussion on this. 
- What information is needed from `searchActivityCards` to grab the exact file needed for preview?
- Please please ask us if you need clarification on what should be done for this ticket. We don't want you to assume things and make it harder for yourself. 

## Resources

Definitely think out your strategy first before writing any code. Talk to @harmitgoswami since you are both doing activity card searching. You can likely share functions.

### PR DESCRIPTION
**Pull request by MinGi-K**
# Description

Added searching functionality and preview functionality to the AddActivity Screen. 

Screen Shot 2023-03-28 at 5 04 24 PM
Screen Shot 2023-03-28 at 5 08 05 PM
Screen Shot 2023-03-28 at 5 05 26 PM
Screen Shot 2023-03-28 at 5 04 39 PM

Ticket # 82

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation changes 
- [ ] Style (formatting, missing semi colons, etc; no production code change)
- [ ] Refactor (refactoring production code, eg. renaming a variable)
- [ ] Chore (updating grunt tasks etc; no production code change)
- [ ] Build (changing build dependencies, updating packages, etc.)

# Checklist:

- [X] My PR title is formatted as [SIS-\<issue number\>] \<issue name\> (eg. [SIS-010] Create UI button)
- [X] I have linked my issue/ticket on the project board to this PR
- [ ] I have sent my PR to #team-sistema-prs on Slack and requested two people to review
- [ ] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if required)
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by yuemily24** [comment]
Hi, there's been a front-end change the designers made. Instead of the tags being lined up, we're going to make a carousel so that users can horizontally scroll through the tags! Please see Figma [here](https://www.figma.com/file/kSZduD0A5hk8sHmvqiC7DE/Sistema-Hi-Fidelity-Designs-%2B-Design-System?node-id=640-15935&t=2FFcHpfq8dezG3cD-0). If you can have this done before Sunday, that'd be great, thank you :) 

The bottom black bar is a scrollbar that progresses as it scrolls, HOWEVER, if it is too much effort to implement (like there isn't a default one in `ScrollView` or something), please don't worry about it. In which case, you can just leave the thin grey bar as a border.

image

**Comment by ramyzhang** [comment]
UR CRAZYYYY MINGIIIIII!!!!!!

File: src/editor/components/LessonSection.js [hunk 1]
```diff
@@ -53,8 +56,8 @@ const LessonSection = ({ sectionType, subtitle, navigation }) => {
         />
 
         {/* Stack of content already inserted, available for further editing/removing */}
-        {store.getState(sectionType).lessonPlan[sectionType].map((arr, i) => {
-          if (arr.content.length > 0) {
+        {lessonPlanState[sectionType].map((arr, i) => {
```
Comment (Lemiex) [inline src/editor/components/LessonSection.js:51]: this is very concise! good idea

File: src/editor/screens/AddActivityCard.js [hunk 2]
```diff
@@ -35,66 +101,257 @@ const AddActivityCard = function ({ navigation, route }) {
...
+    return () => {
+      console.log('Keyboard is unmounted.');
+      keyboardDidShowListener.remove();
+      keyboardDidHideListener.remove();
+    };
+  }, []);
+
+  //onPress function for add Card button
+  const addCard = async () => {
+    console.log('SKFHDSKFHSJKDHFKJSHDFKJSDHKFJHSJDFHSKJDHFJKSDHFK');
```
Comment (Lemiex) [inline src/editor/screens/AddActivityCard.js]: HAhah you might wanna get rid of this

**Review by Lemiex — APPROVED** [review]
great work! what a fat and completed ticket :D

File: src/editor/components/SearchResults.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+const SearchResults = ({
+  name,
+  id,
+  setPreviewInfo,
+  setHighlightedID,
+  isHighlighted,
+}) => {
+  const display_preview = () => {
+    const baseQuery = 'https://www.googleapis.com/drive/v3/files/';
```
Comment (harmitgoswami) [inline src/editor/components/SearchResults.js]: you can use DRIVE_API_URLS.SEARCH_FILES (from config.json in the services folder) here instead, for consistency

File: src/editor/screens/AddActivityCard.js [hunk 4]
```diff
@@ -1,29 +1,95 @@
...
+      } else {
+        const nameSearchTerm = `and name contains '${searchQuery}'`;
+        const tagSearchTerm = TAGS.map((tag, index) => {
+          if (activeTags[index]) {
+            return ` and fullText contains '${tag}'`;
+          }
+        }).join('');
+        const ACTVTTerm = " and fullText contains 'ACTVT'";
+        axios
+          .get('https://www.googleapis.com/drive/v3/files?', {
```
Comment (harmitgoswami) [inline src/editor/screens/AddActivityCard.js]: Same comment as above, use the parameter in config.json

File: src/services/routes/Local.js [hunk 5]
```diff
@@ -53,8 +52,16 @@ export async function readFile(filepath) {
  * @param {String} filepath Full file path to delete
  * @param {String} content Content to write to given file path
  */
-export async function writeFile(filepath, content) {
-  return RNFS.writeFile(filepath, content, 'utf8')
+export async function writeFile(isImage, filepath, content) {
```
Comment (harmitgoswami) [inline src/services/routes/Local.js:55]: just as a heads up: i've changed this in my code, both in terms of conciseness and adding an extra term to deal with image encoding

**Review by harmitgoswami — APPROVED** [review]
LGTM!!!! only some minor changes here and there regarding form, but otherwise YOU DID SUCH A GOOD JOB NICE ONE!!!!111|!!|!!1!!!

File: src/editor/components/SearchResults.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        });
+        console.log(response.data.thumbnailLink);
+      });
+  };
+
+  return (
+    <TouchableOpacity
+      onPress={display_preview}
+      style={[styles.container, isHighlighted && styles.containerHighlighted]}>
+      <Text style={styles.textStyle}>{name || ''}</Text>
```
Comment (yuemily24) [inline src/editor/components/SearchResults.js:37]: Is this supposed to be a nullish operator? 

> [When should I use ?? (nullish coalescing) vs || (logical OR)?](https://stackoverflow.com/questions/61480993/when-should-i-use-nullish-coalescing-vs-logical-or)

File: src/editor/components/NoCardsFound.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  },
+  text2: {
+    fontColor: 'black',
+    fontSize: 15,
+  },
+  text3: {
+    color: '#0078E8',
+    fontSize: 15,
+    fontFamily: 'Mulish-Regular',
+  },
```
Comment (yuemily24) [inline src/editor/components/NoCardsFound.js:64]: You can use `Styles.config.js` for these

File: src/editor/components/NoCardsFound.js
Comment (yuemily24) [inline src/editor/components/NoCardsFound.js:1]: Use react-native-size-matters

File: src/editor/components/SearchResults.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      })
+      .then(function (response) {
+        setHighlightedID(id);
+        setPreviewInfo({
+          url: response.data.thumbnailLink,
+          name: name,
+          id: id,
+        });
+        console.log(response.data.thumbnailLink);
+      });
```
Comment (yuemily24) [inline src/editor/components/SearchResults.js:30]: Convert to arrow func, remove console log

File: src/editor/components/SearchResults.js
Comment (yuemily24) [inline src/editor/components/SearchResults.js:1]: Use react-native-size-matters

File: src/editor/components/SearchResults.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    borderColor: '#36ABFF',
+    borderBottomColor: '#36ABFF',
+    borderBottomWidth: 3,
+  },
+  textStyle: {
+    fontFamily: 'Mulish-Regular',
+    width: '100%',
+    fontSize: 17,
+    color: 'black',
+  },
```
Comment (yuemily24) [inline src/editor/components/SearchResults.js:62]: Styles.config.js text styles

File: src/editor/components/Searchbar.js [hunk 10]
```diff
@@ -1,23 +1,80 @@
...
+
+const Searchbar = ({
+  onChangeText,
+  onFocus,
+  activityList,
+  focused,
+  setPreviewInfo,
+  showNoCards,
+}) => {
+  useEffect(() => {}, [activityList]);
```
Comment (yuemily24) [inline src/editor/components/Searchbar.js:24]:
```suggestion
```
Delete since it does nothing

File: src/editor/components/TagCarousel.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+  const handleScroll = event => {
+    const { x } = event.nativeEvent.contentOffset;
+    const { width } = event.nativeEvent.contentSize;
+    const maxScrollX = width;
+    const scrollPercentage = x / maxScrollX;
+    // Animated.event(
+    //     [{nativeEvent: {contentOffset: {x: scrollBarPosition}}}],
+    //     {useNativeDriver: false}
+    //   )
```
Comment (yuemily24) [inline src/editor/components/TagCarousel.js:20]:
```suggestion
```

**Review by yuemily24 — COMMENTED** [review]
I'm going to merge this because it's currently blocking @Lemiex and all the comments I made are negligible changes that don't affect functionality. 

I left comments so that whoever comes by later to clean stuff up before user testing knows what to address. 

Great job Mingi!

**Pull request merged**